### PR TITLE
feat: add workflow/DAG job dependencies (#312)

### LINKED ISSUE #312: feat: Add workflow/DAG job dependencies
## Summary

Support directed acyclic graph (DAG) based job dependencies, enabling workflow orchestration where jobs run conditionally based on other jobs' completion.

## Motivation

Found in [netresearch/ofelia](https://github.com/netresearch/ofelia) during dependent analysis. Ofelia implements a complete workflow system:
- Job dependencies (prerequisites that must complete first)
- `OnSuccess` / `OnFailure` conditional triggers
- Circular dependency detection via DFS
- Execution tracking per workflow instance

This is also relevant for [weaviate/weaviate](https://github.com/weaviate/weaviate) which manages multiple interdependent background tasks.

## Use Cases

1. **ETL pipelines**: Extract → Transform → Load, each step depends on the previous
2. **Backup workflows**: Dump DB → Compress → Upload to S3 → Cleanup temp files
3. **Multi-stage processing**: Validate → Process → Notify, with error handling at each stage

## Proposed API

```go
// Define jobs
extractID, _ := c.AddFunc("0 2 * * *", extractFunc, cron.WithName("extract"))
transformID, _ := c.AddFunc("@triggered", transformFunc, cron.WithName("transform"))
loadID, _ := c.AddFunc("@triggered", loadFunc, cron.WithName("load"))

// Define dependencies
c.AddDependency(transformID, extractID, cron.OnSuccess)  // transform runs after extract succeeds
c.AddDependency(loadID, transformID, cron.OnSuccess)      // load runs after transform succeeds

// Or declarative:
c.AddWorkflow("etl-pipeline",
    cron.Step("extract", extractFunc, "0 2 * * *"),
    cron.Step("transform", transformFunc).After("extract"),
    cron.Step("load", loadFunc).After("transform"),
)
```

## Complexity

This is a significant feature. Consider:
- Circular dependency detection
- Failure propagation (fail entire workflow vs continue)
- Timeout for entire workflow
- Execution history / audit trail
- Concurrent step execution (fan-out/fan-in)

## Related

- Depends on triggered-only jobs (#311) for non-scheduled steps
- Complements #203 (pause/resume) for workflow control

### PR DESCRIPTION
**Pull request by CybotTM**
## Summary

Adds workflow/DAG job dependencies to go-cron, enabling users to express job dependency graphs with conditional execution: "run B after A succeeds," "run cleanup regardless of outcome," "skip downstream when parent fails."

- **Core types**: `TriggerCondition` (OnSuccess/OnFailure/OnSkipped/OnComplete), `JobResult`, `Dependency`, `WorkflowExecution`
- **Imperative API**: `AddDependency`/`RemoveDependency`/`Dependencies` (+ ByName variants) for programmatic wiring
- **Declarative API**: `NewWorkflow`/`AddWorkflow` builder with `Step`, `StepFunc`, `After`, `Final` for ergonomic multi-step workflows
- **Execution engine**: Channel-based DAG orchestration in the run loop with cascade skip logic, concurrent execution tracking (UUID per run), and configurable retention (`WithWorkflowRetention`)
- **Query API**: `WorkflowStatus`/`ActiveWorkflows` for execution state inspection
- **Context propagation**: `WorkflowExecutionID(ctx)` extracts execution ID from job context
- **Observability**: `OnWorkflowComplete` hook fires when all jobs in an execution resolve
- **Cycle detection**: DFS at add-time prevents cyclic dependencies
- **Atomicity**: `AddWorkflow` validates and registers all steps atomically with rollback on failure

Closes #312

## Test plan

- [x] Unit tests for TriggerCondition, JobResult, Dependency, WorkflowExecution types
- [x] Unit tests for DFS cycle detection (hasCycle + hasCycleByName)
- [x] Integration tests for imperative API (AddDependency/RemoveDependency/Dependencies)
- [x] Integration tests for declarative API (AddWorkflow builder, validation, atomicity)
- [x] Integration tests for execution engine (simple chain, failure cascades, 3-step chain)
- [x] Edge case tests (diamond DAG, paused child, removed entry mid-execution, OnSkipped chain)
- [x] Concurrency test (5 concurrent workflow executions)
- [x] Retention policy test (eviction of oldest completed executions)
- [x] Context propagation test (execution ID shared across workflow jobs)
- [x] Query API tests (WorkflowStatus, ActiveWorkflows)
- [x] Full test suite with race detector: PASS (95.0% coverage)
- [x] golangci-lint: 0 issues
- [x] Pre-push hooks: lint, vulncheck, tests all pass

### CODE REVIEW

File: cron.go [hunk 1]
```diff
@@ -960,6 +1023,24 @@ func (c *Cron) processDueEntries(now time.Time) {
...
+				e.cancelEntryCtx = nil
+				c.removeEntry(e.ID)
+				c.logger.Info("run-once-workflow", "now", now, "entry", e.ID, "removed", true)
+			} else {
+				e.Next = e.Schedule.Next(now)
+				c.hooks.callOnSchedule(e.ID, e.Job, e.Next)
+				c.entries.Update(e)
+				c.logger.Info("run-workflow", "now", now, "entry", e.ID, "next", e.Next)
+			}
+			continue
```
[the remark this answers is not in the document]
Comment (CybotTM) [reply cron.go]: Fixed in 3b16790. Extracted `postDispatchScheduled` and `postDispatchTriggered` helpers that consolidate the runOnce/reschedule logic from both `processDueEntries` and `triggerEntry`.

File: cron.go [hunk 2]
```diff
@@ -1748,6 +2073,237 @@ func (c *Cron) maybeCompactIndexes() {
...
+	c.queryDeps <- req
+	return <-req.reply
+}
+
+// DependenciesByName is the name-based variant of Dependencies.
+func (c *Cron) DependenciesByName(name string) []Dependency {
+	e := c.EntryByName(name)
+	if !e.Valid() {
+		return nil
+	}
```
[the remark this answers is not in the document]
Comment (CybotTM) [reply cron.go]: Fixed in 3b16790. `DependenciesByName` now returns `[]Dependency{}` instead of `nil` for unknown names.

File: cron.go [hunk 3]
```diff
@@ -1748,6 +2073,237 @@ func (c *Cron) maybeCompactIndexes() {
...
+		return ErrCycleDetected
+	}
+	// Idempotent: skip if edge already exists.
+	for _, dep := range c.entryDeps[child] {
+		if dep.ParentID == parent && dep.Condition == condition {
+			return nil
+		}
+	}
+	c.entryDeps[child] = append(c.entryDeps[child], Dependency{ParentID: parent, Condition: condition})
+	c.parentToChildren[parent] = append(c.parentToChildren[parent], child)
```
[the remark this answers is not in the document]
Comment (CybotTM) [reply cron.go]: Good catch! Added a `slices.Contains` guard so each child appears at most once per parent in the reverse index, regardless of how many condition edges exist between them. See commit 3b16790.
